### PR TITLE
perf(auth): avoid unnecessary 30sec TOTP lockout [#376]

### PR DESCRIPTION
**Pull request by rmlibre**
## Description

A rate limit was imposed if a user had previously authenticated with the same TOTP code that they're currently authenticating with. The [TOTP](https://en.wikipedia.org/wiki/Time-based_one-time_password) [spec](https://www.rfc-editor.org/rfc/rfc6238) doesn't disallow repeat codes across time intervals. Such collisions are to be expected when using a pseudo-random function ([HMAC](https://www.rfc-editor.org/rfc/rfc4226)), particularly when using short outputs. The rate limit imposition can harm UX with high authentication latencies without reasonable justification.

## Expected Behavior

A user can authenticate with the same TOTP code if it's a valid code generated at a distinct time interval from the committed TOTP code stored in the database.

## Remediations

1. ``pyotp``'s [timecode](https://github.com/pyauth/pyotp/blob/249674bf2814e1fcc3152c618514159cd44b4610/src/pyotp/totp.py#L106) can be utilized to assess the current time interval.

2. A timecode can be generated at auth initialization, & committed to the database along with the TOTP code upon successful auth.

https://github.com/scidsg/hushline/blob/4eba1e49120871a13f8a36dea1b58c9c5d45844f/hushline/routes.py#L317-L320

https://github.com/scidsg/hushline/blob/4eba1e49120871a13f8a36dea1b58c9c5d45844f/hushline/routes.py#L355-L358

3. Subsequent auth attempts distinguish between time intervals when assessing problematic TOTP code reuse.

https://github.com/scidsg/hushline/blob/4eba1e49120871a13f8a36dea1b58c9c5d45844f/hushline/routes.py#L331-L339

### Passing Workflows:

- [x] [Run Tests](https://github.com/rmlibre/hushline/actions/runs/9654647542)
- [x] [Publish container image](https://github.com/rmlibre/hushline/actions/runs/9654647560)

### CODE REVIEW

**Review by micahflee — CHANGES_REQUESTED** [review]
Thank you for contributing this! The logic looks sound to me, but you added a `timecode` column to the `AuthenticationLog` model and didn't write a new migration for it.

Can you add a migration? You can create a migration by running something like this:

```
source dev_env.sh && poetry run flask db revision -m 'add timecode to AuthenticationLog' --autogenerate
```

**Comment by rmlibre** [comment]
@micahflee 

Commit ([a0555fa][commit-a0555fa]) includes the requested migration.

``build(db): run migrations with `AuthenticationLog.timecode` addition [#377]``

https://github.com/scidsg/hushline/blob/a0555facf789320f66bafa6c26ef9fc53421d600/migrations/versions/568f77aefcb4_add_timecode_to_authenticationlog.py#L1-L7




### Passing Workflows:

- [x] [Run Tests][run-tests-runner]

- [x] [Publish container image][publish-container-image-runner]




### Steps Required on Ubuntu 20.04 LTS:

To work around an error (``ModuleNotFoundError: No module named '_sqlite3'``)<sup>([ref][so-answer-0])</sup> I had to build python3.12 with some extra settings.

**After downloading a Python3.12 interpreter, prepare an altinstall:**

```shell
:~/$  sudo apt-get --fix-missing --fix-broken --install-recommends install wget curl make build-essential libsqlite3-dev zlib1g-dev libncurses5-dev libgdbm-dev libnss3-dev libssl-dev libreadline-dev libffi-dev xz-utils
:~/$  cd <path>/Python-3.12.4
:~/$  ./configure --enable-loadable-sqlite-extensions
:~/$  sudo make
:~/$  sudo make altinstall
```


**Install hushline requirements.**

```shell
:~/$  cd <path>/hushline
:~/hushline$  /usr/local/bin/python3.12 -m venv .env
:~/hushline$  source .env/bin/activate
(.env) :~/hushline$  source dev_env.sh
(.env) :~/hushline$  poetry install
```


Due to another error (``ERROR [flask_migrate] Error: Target database is not up to date.``)<sup>([ref][so-answer-1])</sup> I discovered the db references needed to be updated.

**Make db heads point to current state.**

```shell
(.env) :~/hushline$  poetry run flask db heads
(.env) :~/hushline$  poetry run flask db stamp heads
(.env) :~/hushline$  poetry run flask db current
```


**Make the migration.**

```shell
(.env) :~/hushline$  poetry run flask db revision -m 'add timecode to AuthenticationLog' --autogenerate
```




[//]: # (pr-links)

[commit-a0555fa]: https://github.com/scidsg/hushline/pull/377/commits/a0555facf789320f66bafa6c26ef9fc53421d600

[run-tests-runner]: https://github.com/rmlibre/hushline/actions/runs/9671747636

[publish-container-image-runner]: https://github.com/rmlibre/hushline/actions/runs/9671747637




[//]: # (annotations)

[so-answer-0]: https://stackoverflow.com/a/24449632

[so-answer-1]: https://stackoverflow.com/a/58586211

File: migrations/versions/568f77aefcb4_add_timecode_to_authenticationlog.py [hunk 1]
```diff
@@ -0,0 +1,85 @@
...
+    )
+    op.create_table('message',
+    sa.Column('id', sa.Integer(), nullable=False),
+    sa.Column('content', sa.Text(), nullable=False),
+    sa.Column('user_id', sa.Integer(), nullable=False),
+    sa.Column('secondary_user_id', sa.Integer(), nullable=True),
+    sa.ForeignKeyConstraint(['secondary_user_id'], ['secondary_usernames.id'], ),
+    sa.ForeignKeyConstraint(['user_id'], ['users.id'], ),
+    sa.PrimaryKeyConstraint('id')
+    )
```
Comment (jeremywmoore) [inline migrations/versions/568f77aefcb4_add_timecode_to_authenticationlog.py:74]:
```suggestion
    
    op.add_column('authentication_logs', sa.Column('timecode', sa.Integer(), nullable=True))
```
You don't need to recreate the entire schema with this migration, you just need to add the column to the existing table

File: migrations/versions/568f77aefcb4_add_timecode_to_authenticationlog.py [hunk 2]
```diff
@@ -0,0 +1,85 @@
...
+    # ### end Alembic commands ###
+
+
+def downgrade():
+    # ### commands auto generated by Alembic - please adjust! ###
+    op.drop_table('message')
+    op.drop_table('secondary_usernames')
+    op.drop_table('authentication_logs')
+    op.drop_table('users')
+    op.drop_table('invite_code')
```
Comment (jeremywmoore) [inline migrations/versions/568f77aefcb4_add_timecode_to_authenticationlog.py:84]:
```suggestion
    op.drop_column('authentication_logs', 'timecode')
```

**Review by micahflee — APPROVED** [review]

**Pull request merged**